File: plugins/agentic-sdlc/src/agentic_sdlc/commands/workflows.py
"""Workflows command handler."""

from __future__ import annotations

import sys
from pathlib import Path
from typing import TYPE_CHECKING

import yaml

if TYPE_CHECKING:
    from argparse import Namespace


def get_bundled_workflows_dir() -> Path:
    """Get the directory containing bundled workflow templates."""
    return Path(__file__).parent.parent / "workflows"


def get_user_workflows_dir() -> Path:
    """Get the user-global workflows directory."""
    from os import environ

    if sys.platform == "win32":
        base_dir = Path(environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
    else:
        base_dir = Path(environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))

    return base_dir / "agentic-sdlc" / "workflows"


def get_project_workflows_dir() -> Path:
    """Get the project-local workflows directory."""
    return Path.cwd() / "agentic" / "workflows"


def get_workflow_metadata(path: Path) -> dict:
    """Extract metadata from a workflow YAML file.

    Returns:
        Dict with name, description, and variables from the workflow.
    """
    try:
        with open(path, encoding="utf-8") as f:
            content = yaml.safe_load(f)
            if not isinstance(content, dict):
                return {}
            return {
                "name": content.get("name", path.stem),
                "description": content.get("description", ""),
                "variables": content.get("variables", []),
            }
    except Exception:
        return {"name": path.stem, "description": "", "variables": []}


def list_all_workflows() -> list[tuple[str, Path, str, dict]]:
    """List all available workflows with metadata.

    Returns:
        List of tuples: (workflow_name, path, location_type, metadata)
    """
    workflows: list[tuple[str, Path, str, dict]] = []

    search_locations = [
        (get_project_workflows_dir(), "project"),
        (get_user_workflows_dir(), "user"),
        (get_bundled_workflows_dir(), "bundled"),
    ]

    for directory, location_type in search_locations:
        if directory.exists():
            for workflow_path in sorted(directory.glob("*.yaml")):
                metadata = get_workflow_metadata(workflow_path)
                workflows.append((workflow_path.stem, workflow_path, location_type, metadata))

    return workflows


def cmd_workflows(args: Namespace) -> None:
    """List available workflows with descriptions."""
    workflows = list_all_workflows()

    if not workflows:
        print("No workflows found.")
        print("\nSearched locations:")
        print(f"  - Project: {get_project_workflows_dir()}")
        print(f"  - User:    {get_user_workflows_dir()}")
        print(f"  - Bundled: {get_bundled_workflows_dir()}")
        print("\nUse 'agentic-sdlc init' to copy bundled workflows locally.")
        return

    # Check for --verbose flag
    verbose = hasattr(args, "verbose") and args.verbose

    # Group by location
    by_location: dict[str, list[tuple[str, Path, dict]]] = {}
    for name, path, location, metadata in workflows:
        if location not in by_location:
            by_location[location] = []
        by_location[location].append((name, path, metadata))

    print("Available Workflows")
    print("=" * 50)
    print()

    # Display grouped results
    location_order = ["project", "user", "bundled"]
    location_labels = {"project": "Project", "user": "User", "bundled": "Bundled"}

    for location in location_order:
        if location not in by_location:
            continue

        print(f"{location_labels[location]}:")
        for name, _path, metadata in by_location[location]:
            desc = metadata.get("description", "")
            if desc:
                # Truncate long descriptions
                if len(desc) > 60 and not verbose:
                    desc = desc[:57] + "..."
                print(f"  {name:<25} {desc}")
            else:
                print(f"  {name}")

            # Show variables in verbose mode
            if verbose and metadata.get("variables"):
                vars_list = metadata["variables"]
                required_vars = [v["name"] for v in vars_list if v.get("required", False)]
                optional_vars = [v["name"] for v in vars_list if not v.get("required", False)]
                if required_vars:
                    print(f"    Required: {', '.join(required_vars)}")
                if optional_vars:
                    print(f"    Optional: {', '.join(optional_vars)}")

        print()

    print(f"Total: {len(workflows)} workflow(s)")
    print()
    print("Usage:")
    print("  agentic-sdlc run <workflow-name>")
    print("  agentic-sdlc run <workflow-name> --var key=value")
    print()
    print("Examples:")
    print("  agentic-sdlc run one-shot --var task=\"Add login button\"")
    print("  agentic-sdlc run analyze-single --var analysis_type=bug")
    print("  agentic-sdlc run plan-build-validate --var task=\"Refactor auth\"")

File: src/commands/workflows.ts
/** Workflows command handler. */

import { existsSync, readFileSync, readdirSync } from "node:fs";
import path from "node:path";
import yaml from "js-yaml";

import { getBundledWorkflowsDir, getProjectWorkflowsDir, getUserWorkflowsDir } from "./run.js";

interface WorkflowMetadata {
	name: string;
	description: string;
	variables: {
		name: string;
		type?: string;
		required?: boolean;
		default?: unknown;
		description?: string;
	}[];
}

function getWorkflowMetadata(filePath: string): WorkflowMetadata {
	try {
		const content = yaml.load(readFileSync(filePath, "utf-8")) as Record<string, unknown> | null;
		if (!content || typeof content !== "object") {
			return { name: path.basename(filePath, ".yaml"), description: "", variables: [] };
		}
		return {
			name: (content.name as string) ?? path.basename(filePath, ".yaml"),
			description: (content.description as string) ?? "",
			variables: (content.variables as WorkflowMetadata["variables"]) ?? [],
		};
	} catch {
		return { name: path.basename(filePath, ".yaml"), description: "", variables: [] };
	}
}

function listAllWorkflows(): [string, string, string, WorkflowMetadata][] {
	const workflows: [string, string, string, WorkflowMetadata][] = [];

	const searchLocations: [string, string][] = [
		[getProjectWorkflowsDir(), "project"],
		[getUserWorkflowsDir(), "user"],
		[getBundledWorkflowsDir(), "bundled"],
	];

	for (const [directory, locationType] of searchLocations) {
		if (existsSync(directory)) {
			const files = readdirSync(directory)
				.filter((f) => f.endsWith(".yaml"))
				.sort();
			for (const file of files) {
				const fullPath = path.join(directory, file);
				const metadata = getWorkflowMetadata(fullPath);
				workflows.push([path.basename(file, ".yaml"), fullPath, locationType, metadata]);
			}
		}
	}

	return workflows;
}

export function cmdWorkflows(options: { verbose?: boolean }): void {
	const workflows = listAllWorkflows();

	if (workflows.length === 0) {
		process.stdout.write("No workflows found.\n");
		process.stdout.write("\nSearched locations:\n");
		process.stdout.write(`  - Project: ${getProjectWorkflowsDir()}\n`);
		process.stdout.write(`  - User:    ${getUserWorkflowsDir()}\n`);
		process.stdout.write(`  - Bundled: ${getBundledWorkflowsDir()}\n`);
		process.stdout.write("\nUse 'agentic-forge init' to copy bundled workflows locally.\n");
		return;
	}

	const verbose = options.verbose ?? false;

	// Group by location
	const byLocation: Record<string, [string, string, WorkflowMetadata][]> = {};
	for (const [name, wfPath, location, metadata] of workflows) {
		if (!byLocation[location]) {
			byLocation[location] = [];
		}
		byLocation[location].push([name, wfPath, metadata]);
	}

	process.stdout.write("Available Workflows\n");
	process.stdout.write(`${"=".repeat(50)}\n\n`);

	const locationOrder = ["project", "user", "bundled"];
	const locationLabels: Record<string, string> = {
		project: "Project",
		user: "User",
		bundled: "Bundled",
	};

	for (const location of locationOrder) {
		if (!byLocation[location]) {
			continue;
		}

		process.stdout.write(`${locationLabels[location]}:\n`);
		for (const [name, , metadata] of byLocation[location]) {
			let desc = metadata.description;
			if (desc) {
				if (desc.length > 60 && !verbose) {
					desc = `${desc.slice(0, 57)}...`;
				}
				process.stdout.write(`  ${name.padEnd(25)} ${desc}\n`);
			} else {
				process.stdout.write(`  ${name}\n`);
			}

			if (verbose && metadata.variables.length > 0) {
				process.stdout.write("    Variables:\n");
				for (const v of metadata.variables) {
					const req = v.required ? "required" : "optional";
					const def = v.default !== undefined ? `, default: ${v.default}` : "";
					const typ = v.type ? ` (${v.type})` : "";
					process.stdout.write(`      ${v.name}${typ} [${req}${def}]\n`);
					if (v.description) {
						process.stdout.write(`        ${v.description}\n`);
					}
				}
			}
		}
		process.stdout.write("\n");
	}

	process.stdout.write(`Total: ${workflows.length} workflow(s)\n\n`);
	process.stdout.write("Usage:\n");
	process.stdout.write("  agentic-forge run <workflow> key=value [key=value ...]\n");
	process.stdout.write("  agentic-forge run <workflow> --var key=value\n\n");
	process.stdout.write("Examples:\n");
	process.stdout.write('  agentic-forge run one-shot task="Add login button"\n');
	process.stdout.write("  agentic-forge run analyze-single analysis_type=bug\n");
	process.stdout.write('  agentic-forge run plan-build-review task="Refactor auth"\n');
}
